modules/apps/frontend-view-state/frontend-view-state-service/src/main/java/com/liferay/frontend/view/state/internal/jaxrs/application/FrontendViewStateApplication.java

```diff
@@ -0,0 +1,147 @@
+
+import com.liferay.frontend.view.state.model.FVSActiveEntry;
+import com.liferay.frontend.view.state.model.FVSEntry;
+import com.liferay.frontend.view.state.service.FVSActiveEntryLocalService;
+import com.liferay.frontend.view.state.service.FVSEntryLocalService;
+import com.liferay.portal.kernel.json.JSONFactory;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.model.User;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Application;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.jaxrs.whiteboard.JaxrsWhiteboardConstants;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+@Component(
+	property = {
+		JaxrsWhiteboardConstants.JAX_RS_APPLICATION_BASE + "=/frontend-view-state/app",
+		JaxrsWhiteboardConstants.JAX_RS_NAME + "=Liferay.Frontend.View.State",
+		"auth.verifier.auth.verifier.PortalSessionAuthVerifier.urls.includes=/*",
+		"auth.verifier.guest.allowed=true", "liferay.oauth2=false"
+	},
+	service = Application.class
+)
+public class FrontendViewStateApplication {
+
+	@GET
+	@Path("/active-view/{clayDataSetDisplayId}")
+	@Produces(MediaType.APPLICATION_JSON)
+	public Response getActiveView(
+		@PathParam("clayDataSetDisplayId") String clayDataSetDisplayId,
+		@QueryParam("plid") long plid,
+		@QueryParam("portletId") String portletId, @Context User user) {
+
+		try {
+			FVSActiveEntry fvsActiveEntry =
+				_fvsActiveEntryLocalService.fetchFVSActiveEntry(
+					user.getUserId(), clayDataSetDisplayId, plid, portletId);
+
+			if (fvsActiveEntry == null) {
+				return Response.status(
+					Response.Status.NOT_FOUND
+				).build();
+			}
+
+			FVSEntry fvsEntry = _fvsEntryLocalService.getFVSEntry(
+				fvsActiveEntry.getFvsEntryId());
+
+			return Response.ok(
+				_jsonFactory.createJSONObject(fvsEntry.getViewState()),
+				MediaType.APPLICATION_JSON
+			).build();
+		}
+		catch (Exception exception) {
+			_log.error("Unable to get active view", exception);
+		}
+
+		return Response.status(
+			Response.Status.NOT_FOUND
+		).build();
+	}
+
+	@Consumes(MediaType.APPLICATION_JSON)
+	@Path("/active-view/{clayDataSetDisplayId}/save")
+	@POST
+	public Response saveActiveView(
+		@PathParam("clayDataSetDisplayId") String clayDataSetDisplayId,
+		@QueryParam("plid") long plid,
+		@QueryParam("portletId") String portletId, @Context User user,
+		String viewState) {
+
+		try {
+			FVSActiveEntry fvsActiveEntry =
+				_fvsActiveEntryLocalService.fetchFVSActiveEntry(
+					user.getUserId(), clayDataSetDisplayId, plid, portletId);
+
+			if (fvsActiveEntry == null) {
```

Comment:
Checking the logic, it seems like an UPSERT to me. Creating the resource if it not exists or update it otherwise. We are modeling this kind of operation as a PUT, don't know if we want to follow the same convention here or it is to help the frontend calling the method from a form or something like that.

The PUT also makes sense to me to get rid of the "/save" chink in the path and refer directly to the resource to better follow the REST guidelines.

Reply:
I really though that was the most RESTful way to do it, but because this was copied from an already existing API I did it this way. One concern, however, is that the frontend side (JS) doesn't now if it is and update or a creation. But I guess that scenario is  covered by PUT, too (if you PUT something that already exists, you simply overwrite it, right?).

Also, I don't know if there is any consideration we should make in regards to the clay components invoking this API. I mean, I don't know how difficult/easy it is for them to do PUT vs POST...

Reply:
This is a very good feedback, PUT is meant to be used in upsert operations as the standard indicates: "If the Request-URI does not point to an existing resource (...) the origin server can create the resource with that URI." and you can pass a `fetch` in Clay so you should easily be able to do a `PUT`. I would change this in the future (if the PR passes).